plugins/crd/crd_utils.go

```diff
@@ -46,31 +54,66 @@ func newCustomResourceDefinition(plugin *Plugin, FullName, Group, Version, Plura
 		GetOpenAPIDefinitions: v1.GetOpenAPIDefinitions,
 	})
 
-	crdClient := plugin.apiclientset.ApiextensionsV1beta1().CustomResourceDefinitions()
+	crdClient := plugin.apiclientset
+	// Add NSM CRD version
+	crd.ObjectMeta.Annotations = map[string]string{
+		nsmCRDVersionAnnotationKey: nsmCRDVersion,
+	}
+	if err := createCRDObject(crd, crdClient); err != nil {
+		plugin.Log.Errorf("fail to create CRD with error: %#v", err)
+		return err
+	}
+
+	return nil
+}
 
+func createCRDObject(newCRD *apiextv1beta1.CustomResourceDefinition, crdClient *apiextcs.Clientset) error {
 	// First check if the CRD already exists
-	oldCRD, err := crdClient.Get(crd.Name, metav1.GetOptions{})
+	oldCRD, err := crdClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(newCRD.Name, metav1.GetOptions{})
 	if err != nil && !apierrors.IsNotFound(err) {
-		plugin.Log.Errorf("error getting CRD %s, type %s", crd.Name, crd.Spec.Names.Kind)
-		return err
+		return fmt.Errorf("error getting CRD %s, type %s", newCRD.Name, newCRD.Spec.Names.Kind)
 	}
 	if apierrors.IsNotFound(err) {
 		// If the CRD does not exist, try to create it
-		if _, err := crdClient.Create(crd); err != nil {
-			plugin.Log.Errorf("error creating CRD %s, type %s", crd.Name, crd.Spec.Names.Kind)
-			return err
+		if _, err := crdClient.ApiextensionsV1beta1().CustomResourceDefinitions().Create(newCRD); err != nil {
+			return fmt.Errorf("fail creating CRD %s, type %s with error: %#v", newCRD.Name, newCRD.Spec.Names.Kind, err)
 		}
-		plugin.Log.Infof("created CRD %s, type %s", crd.Name, crd.Spec.Names.Kind)
+		return nil
 	}
-	if err == nil {
-		// Now we try to update the CRD
-		crd.ResourceVersion = oldCRD.ResourceVersion
-		if _, err := crdClient.Update(crd); err != nil {
-			plugin.Log.Errorf("error updating CRD %s, type %s", crd.Name, crd.Spec.Names.Kind)
-			return err
-		}
-		plugin.Log.Infof("updated CRD %s, type %s", crd.Name, crd.Spec.Names.Kind)
+	// Check if CRD has the version annotation, if not, then it means it is old CRD
+	// and we update it uncoditionally.
+	version, ok := oldCRD.ObjectMeta.Annotations[nsmCRDVersionAnnotationKey]
+	if !ok {
+		// Exisiting CRD does not have the version annotation, updating it to new definition
+		// uncoditionally
+		return updateCRD(newCRD, oldCRD.ResourceVersion, crdClient)
+	}
+	// Existing CRD has version info, next check is to see if existing CRD version is "<" or "==" or ">"
+	// if "<" than new CRD, it will be updated, if "==", then no action , if ">" than new CRD, then we will fail as
+	// it possible, that older version of NSM controller started on a cluster with newer CRD definitions.
```

Comment:
We may want to eventually consider downgrades of NSM, but for now, I think logic is fine.